src/V8JsiRuntime.cpp

```diff
@@ -593,15 +624,21 @@ V8Runtime::V8Runtime(V8RuntimeArgs &&args) : args_(std::move(args)) {
 
   v8::Isolate::Scope isolate_scope(isolate_);
   v8::HandleScope handleScope(isolate_);
-  context_.Reset(GetIsolate(), CreateContext(isolate_));
 
-  v8::Context::Scope context_scope(context_.Get(GetIsolate()));
+  auto context = CreateContext(isolate_);
+  context_.Reset(isolate_, context);
+
+  // Associate the Runtime with the V8 context
+  context->SetAlignedPointerInEmbedderData(ContextEmbedderIndex::Runtime, this);
+  // Used by Environment::GetCurrent to know that we are on a node context.
+  context->SetAlignedPointerInEmbedderData(
+      ContextEmbedderIndex::ContextTag, V8Runtime::RuntimeContextTagPtr);
+
+  v8::Context::Scope context_scope(context);
 

```

Comment:
We ideally want to allow multiple contexts in the same thread sharing the same isolate .. This will break it.

Reply:
Could you explain how it breaks it? This code associates a context with the Runtime. So, we can always retrieve the runtime from the context. This code does not do anything related to a thread.
The thread affinity is introduced by the TLS variables that we have and are continuing to add. The TLS variables must be removed in future and replaced with this mechanism where we can easily get from the Runtime to context and vice versa.
The next step will be to have APIs such as napi_ext_open_env_scope/napi_ext_close_env_scope that bind the Runtime to a specific thread while the Runtime is in scope.

Reply:
Ah .. I see.. When reading this last time, i thought this scope object is kept with the runtime .. Here, you are only creating a stack variable.. Im curious to konw what is the purpose of this statement ? No new code seems to be added in the constructor after this statement.

---
In reply to: [611886893](https://github.com/microsoft/v8-jsi/pull/51#discussion_r611886893) [](ancestors = 611886893)

Reply:
Please note that I did not add this statement: it is the previously existing code `v8::Context::Scope context_scope(context_.Get(GetIsolate()));` - it was simplified to use the `context` variable. If I remove it, then tests continue to work, but I want to keep it for now because I do not know all existing scenarios.

---
In reply to: [611890075](https://github.com/microsoft/v8-jsi/pull/51#discussion_r611890075) [](ancestors = 611890075,611886893)